enhancements/cert-manager/istio-csr-controller.md

```diff
@@ -0,0 +1,929 @@
+
+	// IstiodPrivateKeySize is the key size to be for RSAKey.
+	// +kubebuilder:default:=2048
+	// +optional
+	IstiodPrivateKeySize int32 `json:"istiodPrivateKeySize,omitempty"`
+}
+
+// ServerConfig is for configuring the server endpoint used by istio
+// for ontaining the certificates.
+type ServerConfig struct {
+	// Address to serve istio-csr gRPC service.
+	// +kubebuilder:default:="0.0.0.0"
+	// +optional
+	Address string `json:"address,omitempty"`
+
+	// Port to serve istio-csr gRPC service.
+	// +kubebuilder:default:="6443"
+	// +optional
+	Port string `json:"port,omitempty"`
+
+	// CertificateKeySize is the server's serving certificate's key size.
+	// +kubebuilder:default:=2048
+	// +optional
+	CertificateKeySize int32 `json:"certificateKeySize,omitempty"`
+
+	// SignatureAlgorithm is the signature algorithm to use when generating
+	// private keys. At present only RSA and ECDSA are supported.
+	// +kubebuilder:default:="RSA"
+	// +kubebuilder:validation:Enum:="RSA";"ECDSA"
+	// +optional
+	SignatureAlgorithm string `json:"signatureAlgorithm,omitempty"`
+
+	// MaxCertificateDuration is the maximum validity duration that can be
+	// requested for a certificate.
+	// +kubebuilder:default:="1h"
+	// +optional
+	MaxCertificateDuration time.Duration `json:"certificateDuration,omitempty"`
+}
+
+// IstioConfig is for configuring the istio specifics.
+type IstioConfig struct {
+	// Revisions are the istio revisions that are currently installed in the cluster.
+	// Changing this field will modify the DNS names that will be requested for
+	// the istiod certificate.
+	// +kubebuilder:default:=["basic"]
+	// +optional
+	Revisions []string `json:"revisions,omitempty"`
+
+	// namespace of the istio control-plane. In the same namespace issuer will be created
+	// used for obtaining the serving certificates.
+	Namespace string `json:"namespace,omitempty"`
+}
+
+// ControllerConfig is for configuring the controller for setting up
+// defaults to enable istio-csr agent.
+type ControllerConfig struct {
+	// labels to apply to all resources created for istio-csr agent deployment.
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// issuerConfig is for setting the issuer specifics to be created by the controller.
+	IssuerConfig *IssuerConfig `json:"issuerConfig,omitempty"`
+}
+
+type IssuerConfig struct {
+	// issuerType is for specifying issuer type `ClusterIssuer` or `Issuer`.
+	Issuertype string `json:"issuerType,omitempty"`
+
+	// duration is for setting the default certificate validity duration.
+	Duration string `json:"duration,omitempty"`
+
+	// renewBefore is for specifying the certificate renewal time before the certificate expiry.
+	RenewBefore string `json:"renewBefore,omitempty"`
+}
+
+// IstioCSRStatus is the most recently observed status of the IstioCSR.
+type IstioCSRStatus struct {
+	// conditions holds information of the current state of the istio-csr agent deployment.
+	Conditions *metav1.Condition `json:"conditions,omitempty"`
```

Comment:
Any status that indicates failure, should set the operator status to degraded disabling upgrades or any further add-on operations.

Reply:
Conditions will indicate degrade status if there is any unrecoverable error.

Reply:
Yes, if any of the required resources are not in expected state, `IstioCSRStatus` condition will be updated with error observed in the problematic resource.

Reply:
Can you explicitly add the statements in the proposal?

Reply:
Also, the scenario where istio-csr CR is removed, the expected operator status should be updated.